services/MerchStoreService.ts

```diff
@@ -431,34 +441,236 @@ export default class MerchStoreService {
+  }
+
+  /**
+   * Marks an order as missed. An order can be marked as missed only if it's previous status was PLACED,
+   * and its associated pickup event has already passed. Only admins can mark orders as missed
+   * @param uuid order uuid
+   * @returns updated order
+   */
+  public async markOrderAsMissed(uuid: Uuid, proxy: UserModel): Promise<OrderModel> {
+    return this.transactions.readWrite(async (txn) => {
+      const orderRespository = Repositories.merchOrder(txn);
+      const order = await orderRespository.findByUuid(uuid);
+      if (!order) throw new NotFoundError('Order not found');
+      if (order.status !== OrderStatus.PLACED) {
+        throw new UserError('Cannot mark an order as missed if it\'s already been cancelled, missed, or fulfilled');
+      }
+      // compare with start date and not end date so that store admins can mark orders
+      // as missed during the event and not necessarily limited to after the event
+      // (e.g. in the case where the event ends early)
+      if (new Date() < moment(order.pickupEvent.start).toDate()) {
+        throw new NotFoundError('Cannot mark an order as missed if its pickup event hasn\'t started yet');
+      }
+      const upsertedOrder = await orderRespository.upsertMerchOrder(order, { status: OrderStatus.PICKUP_MISSED });
+      const orderUpdateInfo = await MerchStoreService
+        .buildOrderUpdateInfo(upsertedOrder, upsertedOrder.pickupEvent, txn);
+      const { user } = order;
+      await this.emailService.sendOrderPickupMissed(user.email, user.firstName, orderUpdateInfo);
+
+      const activityRepository = Repositories.activity(txn);
+      await activityRepository.logActivity({
+        user,
+        type: ActivityType.ORDER_MISSED,
+        description: `Order ${order.uuid} marked as missed for ${user.uuid} by ${proxy.uuid}`,
+      });
+      return upsertedOrder;
+    });
+  }
+
+  /**
+   * Cancels a merch order, refunding the user of its credits if the user is the one who cancelled the order.
+   * @param uuid order uuid
+   * @param user user cancelling the order
+   */
+  public async cancelMerchOrder(uuid: Uuid, user: UserModel): Promise<void> {
+    return this.transactions.readWrite(async (txn) => {
+      const orderRespository = Repositories.merchOrder(txn);
+      const order = await orderRespository.findByUuid(uuid);
+      if (!order) throw new NotFoundError('Order not found');
+      if (!user.isAdmin() && order.user.uuid !== user.uuid) {
+        throw new ForbiddenError('Members cannot cancel other members\' orders');
+      }
+      if (MerchStoreService.isInactiveOrder(order)) throw new UserError('Cannot cancel an inactive order');
+      if (MerchStoreService.isLessThanTwoDaysBeforePickupEvent(order.pickupEvent)) {
+        throw new NotFoundError('Cannot cancel an order with a pickup date less than 2 days away');
+      }
+      const customer = order.user;
```

Comment:
Is this used anywhere?

Reply:
Yeah, it's used in the activity logs to describe who's order has been cancelled (since admins can cancel other members orders in the case of 'cancel all pending orders')